File: AppleVisionPro/VisionPro.cpp
#include <pybind11/embed.h>
#include <iostream>
#include <fstream>
#include <chrono>
#include <thread>

namespace py = pybind11;

int main() {
    // Print a message to verify the program started
    std::cout << "Start file execution..." << std::endl;

    // Start the Python interpreter
    py::scoped_interpreter guard{};

    try {
        std::cout << "1" << std::endl;
        // Import the avp_stream library
        py::module avp_stream = py::module::import("avp_stream");

        // Create an instance of VisionProStreamer
        std::string avp_ip = "10.29.194.158";
        py::object VisionProStreamer = avp_stream.attr("VisionProStreamer");
        std::cout << "1.5" << std::endl;
        py::object streamer = VisionProStreamer(avp_ip, true);

        std::cout << "2" << std::endl;
        // Open a text file for writing
        std::ofstream file("streamed_data.txt");
        if (!file.is_open()) {
            std::cerr << "Failed to open file for writing.\n";
            return 1;
        }
        std::cout << "3" << std::endl;
        // Timing
        auto start_time = std::chrono::steady_clock::now();
        double timestep = 0.005; // 5 ms
        double next_timestep = timestep;
        std::cout << "4" << std::endl;
        while (true) {
            auto current_time = std::chrono::steady_clock::now();
            double relative_timestamp = std::chrono::duration<double>(current_time - start_time).count();
            std::cout << "5" << std::endl;
            if (relative_timestamp >= next_timestep) {
                // Call the Python method to get the latest data
                py::object latest = streamer.attr("latest");
                py::object right_wrist = latest.attr("get")("right_wrist");

                // Convert right wrist data to string
                std::string right_wrist_str = py::str(right_wrist).cast<std::string>();

                // Write to file
                file << "Timestamp: " << next_timestep << "\n";
                file << "Right Wrist: " << right_wrist_str << "\n\n";
                file.flush();

                // Print live data to the console
                std::cout << "Timestamp: " << next_timestep
                          << " | Right Wrist: " << right_wrist_str << std::endl;

                // Update the next timestep
                next_timestep += timestep;
            }

            // Sleep to reduce CPU usage
            std::this_thread::sleep_for(std::chrono::milliseconds(1));
        }

    } catch (const py::error_already_set& e) {
        std::cerr << "Python error: " << e.what() << "\n";
    }

    return 0;
}

File: AppleVisionPro/VisionPro.py
import time
from avp_stream import VisionProStreamer
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D


avp_ip = "10.31.153.17"  # Replace with your actual IP
s = VisionProStreamer(ip=avp_ip, record=True)


# Placeholder for translational data
translations = []


# Open a text file for writing
with open('streamed_data.txt', 'w') as f:
    start_time = time.time()  # Capture the start time
    last_written_time = 0  # Initialize the last written timestamp
    timestep = 0.005  # Desired time interval (5 ms)
    next_timestep = timestep  # Start with the first timestep

    try:
        while True:
            current_time = time.time()
            relative_timestamp = current_time - start_time

            # Check if the current relative timestamp has reached or exceeded the next timestep
            if relative_timestamp >= next_timestep:

                r = s.latest  # Get the latest data

                # Write data to the text file
                f.write(f"Timestamp: {next_timestep:.6f}\n")
                f.write(f"Right Wrist: {r['right_wrist'].tolist()}\n")
                #f.write(f"Right Finger Knuckle: {r['right_fingers'][11].tolist()}\n")
                #f.write(f"Right Finger intermediate base: {r['right_fingers'][12].tolist()}\n")
                #f.write(f"Right Finger tip: {r['right_fingers'][14].tolist()}\n")
                f.write("\n")  # Separate each frame with a blank line

                # Flush to ensure data is written immediately
                f.flush()

                # Update the next timestep
                next_timestep += timestep

            # Sleep for a short duration to avoid busy waiting
            time.sleep(0.001)  # 1 ms sleep to reduce CPU usage

    except KeyboardInterrupt:
        print("Data streaming stopped. Saved to streamed_data.txt")



    # Separate data for plotting
    ids = [t[0] for t in translations]  # Index IDs
    x = [t[1][0] for t in translations]  # X translations
    y = [t[1][1] for t in translations]  # Y translations
    z = [t[1][2] for t in translations]  # Z translations

    # Create a 3D scatter plot
    fig = plt.figure()
    ax = fig.add_subplot(111, projection='3d')

    # Plot points and add text labels for IDs
    ax.scatter(x, y, z, c='blue', marker='o', label="Translation Points")


                # Set labels
    ax.set_xlabel('X')
    ax.set_ylabel('Y')
    ax.set_zlabel('Z')
    ax.set_title('3D Translation Positions with IDs')

    plt.legend()
    plt.show()